src/DataTypes/transformTypesRecursively.cpp

```diff
@@ -8,12 +8,58 @@
 namespace DB
 {
 
-void transformTypesRecursively(DataTypes & types, std::function<void(DataTypes &, const TypeIndexesSet &)> transform_simple_types, std::function<void(DataTypes &, const TypeIndexesSet &)> transform_complex_types)
+TypeIndexesSet getTypesIndexes(const DataTypes & types)
 {
     TypeIndexesSet type_indexes;
     for (const auto & type : types)
         type_indexes.insert(type->getTypeId());
+    return type_indexes;
+}
+
+void transformTypesRecursively(DataTypes & types, std::function<void(DataTypes &, const TypeIndexesSet &)> transform_simple_types, std::function<void(DataTypes &, const TypeIndexesSet &)> transform_complex_types)
+{
+    TypeIndexesSet type_indexes = getTypesIndexes(types);
+
+    /// Nullable
+    if (type_indexes.contains(TypeIndex::Nullable))
```

Comment:
Just moved Nullable processing before other complex types like `Array/Map/Tuple` to properly handle case when we have smth like `[Nullable(Nothing), Array(...)]`